tensorrt_llm/_torch/auto_deploy/transform/library/mxfp4_moe.py

```diff
@@ -0,0 +1,320 @@
+    return float(alpha), float(limit)
+
+
+def _get_topk_from_router(node: Node) -> int:
+    # torch_moe_router(hidden, weight, bias, top_k=2)
+    if "top_k" in node.kwargs:
+        return int(node.kwargs["top_k"])
+    return int(node.args[3]) if len(node.args) >= 4 else 2
+
+
+def _register_mxfp4_expert_params(
+    gm: GraphModule,
+    gate_up_w_name: str,
+    gate_up_b_name: str,
+    down_w_name: str,
+    down_b_name: str,
+) -> Tuple[str, str, str, str]:
+    """Create (if missing) the four MXFP4 params under the experts module and return their full names.
+
+    Returns:
+      (gu_blocks_name, gu_scales_name, dn_blocks_name, dn_scales_name)
+    """
+    # Shapes from existing params
+    gu_b = gm.get_parameter(gate_up_b_name)  # [E, 2I]
+    gu_w = gm.get_parameter(gate_up_w_name)  # [E, 2I, H]
+    dn_b = gm.get_parameter(down_b_name)  # [E, H]
+
+    E = int(gu_b.shape[0])
+    I2 = int(gu_b.shape[1])  # 2I
+    In = I2 // 2
+
+    # infer H from gu_w shape
+    assert gu_w.dim() == 3, "gate_up_w must be rank-3"
+    if gu_w.shape[1] == I2:
+        H = int(gu_w.shape[2])
+    else:
+        # Fallback: use down bias last dim
+        H = int(dn_b.shape[1])
+
+    # Compute block dims (assume divisible; zero-init anyway)
+    H_blk = max(1, H // 32)
+    I_blk = max(1, In // 32)
+
+    experts_mod, experts_path, _ = get_submodule_of_param(gm, gate_up_w_name)
+
+    # New param names under experts module
+    gu_blocks_name = "gate_up_proj_blocks"
+    gu_scales_name = "gate_up_proj_scales"
+    dn_blocks_name = "down_proj_blocks"
+    dn_scales_name = "down_proj_scales"
+
+    # Zero-init tensors (uint8 for blocks/scales)
+    gu_blocks = torch.zeros((E, 2 * In, H_blk, 16), dtype=torch.uint8)
+    gu_scales = torch.zeros((E, 2 * In, H_blk), dtype=torch.uint8)
+    dn_blocks = torch.zeros((E, H, I_blk, 16), dtype=torch.uint8)
+    dn_scales = torch.zeros((E, H, I_blk), dtype=torch.uint8)
+
+    experts_mod.register_parameter(gu_blocks_name, nn.Parameter(gu_blocks, requires_grad=False))
+    experts_mod.register_parameter(gu_scales_name, nn.Parameter(gu_scales, requires_grad=False))
+    experts_mod.register_parameter(dn_blocks_name, nn.Parameter(dn_blocks, requires_grad=False))
+    experts_mod.register_parameter(dn_scales_name, nn.Parameter(dn_scales, requires_grad=False))
+
+    # Full GM attribute paths for new params
+    prefix = (experts_path + ".") if experts_path else ""
+    return (
+        prefix + gu_blocks_name,
+        prefix + gu_scales_name,
+        prefix + dn_blocks_name,
+        prefix + dn_scales_name,
+    )
+
+
+@TransformRegistry.register("quantize_mxfp4_moe")
+class InsertMXFP4MLP(BaseTransform):
+    """
+    Replace (torch_moe_router -> torch_moe_dense_mlp) with a single auto_deploy::triton_mxfp4_moe op,
+    and register MXFP4 expert params (blocks + scales) on the experts module.
+    """
+
+    algo_name: str = "mxfp4"
+
+    def _apply(
```

Comment:
I am curious why you decided to implement this pattern matcher without the pattern matching utility?